nova/virt/vmwareapi/host.py

```diff
@@ -90,45 +91,58 @@ def update_status(self):
-                                                   self._cluster)
+            per_host_stats = vm_util.get_stats_from_cluster_per_host(
+                self._session, self._cluster)
         except (vexc.VimConnectionException, vexc.VimAttributeException) as ex:
             # VimAttributeException is thrown when vpxd service is down
             LOG.warning("Failed to connect with %(node)s. "
                         "Error: %(error)s",
-                        {'node': self._host_name, 'error': ex})
+                        {'node': self._cluster_node_name, 'error': ex})
             self._set_host_enabled(False)
             return data
 
-        data["vcpus"] = stats['cpu']['vcpus']
-        data["vcpus_reserved"] = stats['cpu']['reserved_vcpus']
-        data["disk_total"] = capacity / units.Gi
-        data["disk_available"] = freespace / units.Gi
-        data["disk_used"] = data["disk_total"] - data["disk_available"]
-        data["host_memory_total"] = stats['mem']['total']
-        data["host_memory_free"] = stats['mem']['free']
-        data['host_memory_reserved'] = stats['mem']['reserved_memory_mb']
-        data["hypervisor_type"] = self._hypervisor_type
-        data["hypervisor_version"] = self._hypervisor_version
-        data["hypervisor_hostname"] = self._host_name
-        data["supported_instances"] = [
-            (obj_fields.Architecture.I686,
-             obj_fields.HVType.VMWARE,
-             obj_fields.VMMode.HVM),
-            (obj_fields.Architecture.X86_64,
-             obj_fields.HVType.VMWARE,
-             obj_fields.VMMode.HVM)]
-        data["cpu_model"] = self.to_cpu_model()
+        local_gb = capacity / units.Gi
+        local_gb_used = (capacity - free) / units.Gi
+        local_gb_max_free = max_free / units.Gi
+
+        defaults = {
+            "local_gb": local_gb,
+            "local_gb_used": local_gb_used,
+            "local_gb_max_free": local_gb_max_free,
+            "supported_instances": [
+                (obj_fields.Architecture.I686,
+                 obj_fields.HVType.VMWARE,
+                 obj_fields.VMMode.HVM),
+                (obj_fields.Architecture.X86_64,
+                 obj_fields.HVType.VMWARE,
+                 obj_fields.VMMode.HVM)],
+            "numa_topology": None,
+        }
+
+        for host, stats in six.iteritems(per_host_stats):
+            data[host] = self._merge_stats(host, stats, defaults)
+
+        cluster_stats = vm_util.aggregate_stats_from_cluster(per_host_stats)
+        cluster_stats["hypervisor_type"] = self._hypervisor_type
+        cluster_stats["hypervisor_version"] = self._hypervisor_version
+        data[self._cluster_node_name] = self._merge_stats(
+            self._cluster_node_name, cluster_stats, defaults)
 
         self._stats = data
         if self._auto_service_disabled:
             self._set_host_enabled(True)
         return data
 
+    def _merge_stats(self, host, stats, defaults):
+        result = defaults.copy()
```

Comment:
I don't think a simple `.copy()` is enough. Also at other places handling these dicts. If we have nested lists or dicts, those will stay the same and if edited and will change everywhere (e.g. `supported_instances`).
```python
>>> import copy
>>> a = {'b': [1, 2, 3]}
>>> c = a.copy()
>>> c
{'b': [1, 2, 3]}
>>> a['b'].append('asdf')
>>> a
{'b': [1, 2, 3, 'asdf']}
>>> c
{'b': [1, 2, 3, 'asdf']}
>>> c = copy.deepcopy(a)
>>> c
{'b': [1, 2, 3, 'asdf']}
>>> a
{'b': [1, 2, 3, 'asdf']}
>>> a['b'].append(4)
>>> c
{'b': [1, 2, 3, 'asdf']}
>>> a
{'b': [1, 2, 3, 'asdf', 4]}
```

Reply:
I'll change it to a deep copy